extension/src/ohlc.rs

```diff
@@ -0,0 +1,644 @@
+
+            if other.close.ts > self.close.ts {
+                self.close = other.close;
+            }
+        }
+
+        pub fn open(&self) -> f64 {
+            self.open.val
+        }
+
+        pub fn high(&self) -> f64 {
+            self.high.val
+        }
+
+        pub fn low(&self) -> f64 {
+            self.low.val
+        }
+
+        pub fn close(&self) -> f64 {
+            self.close.val
+        }
+
+        pub fn open_at(&self) -> i64 {
+            self.open.ts
+        }
+
+        pub fn high_at(&self) -> i64 {
+            self.high.ts
+        }
+
+        pub fn low_at(&self) -> i64 {
+            self.low.ts
+        }
+
+        pub fn close_at(&self) -> i64 {
+            self.close.ts
+        }
+    }
+
+    impl Default for OpenHighLowClose<'_> {
+        fn default() -> Self {
+            Self::new()
+        }
+    }
+    ron_inout_funcs!(OpenHighLowClose);
+}
+
+use toolkit_experimental::OpenHighLowClose;
+
+#[pg_extern(immutable, parallel_safe, schema = "toolkit_experimental")]
+pub fn ohlc_transition(
+    state: Internal,
+    ts: Option<crate::raw::TimestampTz>,
+    price: Option<f64>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Internal> {
+    ohlc_transition_inner(unsafe { state.to_inner() }, ts, price, fcinfo).internal()
+}
+pub fn ohlc_transition_inner(
+    state: Option<Inner<OpenHighLowClose>>,
+    ts: Option<crate::raw::TimestampTz>,
+    price: Option<f64>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Inner<OpenHighLowClose>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            let point = match (ts, price) {
+                (_, None) | (None, _) => return state,
+                (Some(ts), Some(val)) => TSPoint { ts: ts.into(), val },
+            };
+            match state {
+                None => {
+                    let mut ohlc = OpenHighLowClose::new();
+                    ohlc.add_value(point);
+                    Some(ohlc.into())
+                }
+                Some(mut ohlc) => {
+                    ohlc.add_value(point);
+                    Some(ohlc)
+                }
+            }
+        })
+    }
+}
+
+#[pg_extern(immutable, parallel_safe, schema = "toolkit_experimental")]
+pub fn ohlc_rollup_trans(
+    state: Internal,
+    value: Option<OpenHighLowClose>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Internal> {
+    ohlc_rollup_trans_inner(unsafe { state.to_inner() }, value, fcinfo).internal()
+}
+pub fn ohlc_rollup_trans_inner<'input>(
+    state: Option<Inner<OpenHighLowClose<'input>>>,
+    value: Option<OpenHighLowClose<'input>>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Inner<OpenHighLowClose<'input>>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || match (state, value) {
+            (state, None) => state,
+            (None, Some(value)) => {
+                let mut state = OpenHighLowClose::new();
+                state.combine(&value);
+                Some(state.into())
```

Comment:
You should be able to just return value or a copy of value here.

Reply:
sure can :)